File: Day19.cpp
#include <chrono>
#include <iostream>
#include <map>
#include <string>
#include <utility>

std::map<int, std::map<std::string, int>> checkpoints;

void runPrint(int &bestGeodes, std::map<std::string, int> &blueprint, int minutes, int maxMinutes, int oreBots, int ore, int clayBots, int clay, int obsBots, int obs, int geodeBots, int geode, int &maxOre, int &maxClay, int &maxObs) {
    int newMinutes = minutes + 1;

    bool oreBuild = ore >= blueprint["ore"] && oreBots < maxOre;
    bool clayBuild = ore >= blueprint["clay"] && clayBots < maxClay;
    bool obsBuild = ore >= blueprint["obs-1"] && clay >= blueprint["obs-2"] && obsBots < maxObs;
    bool geodeBuild = ore >= blueprint["geode-1"] && obs >= blueprint["geode-2"];

    int newOre = ore + oreBots;
    int newClay = clay + clayBots;
    int newObs = obs + obsBots;
    int newGeode = geode + geodeBots;

    if (newMinutes >= maxMinutes) {
        if (newGeode > bestGeodes) {
            bestGeodes = newGeode;
        }
        return;
    }

    int remainder = maxMinutes - newMinutes;
    int maxPossible = newGeode + (remainder * geodeBots) + (remainder * (remainder + 1)) / 2;
    if (maxPossible <= bestGeodes) {
        return;
    }

    if (oreBots <= checkpoints[newMinutes]["oreBots"] && newOre <= checkpoints[newMinutes]["ore"] && clayBots <= checkpoints[newMinutes]["clayBots"] && newClay <= checkpoints[newMinutes]["clay"] && obsBots <= checkpoints[newMinutes]["obsBots"] && newObs <= checkpoints[newMinutes]["obs"] && geodeBots <= checkpoints[newMinutes]["geodeBots"] && newGeode <= checkpoints[newMinutes]["geode"]) {
        if (oreBots != checkpoints[newMinutes]["oreBots"] || newOre != checkpoints[newMinutes]["ore"] || clayBots != checkpoints[newMinutes]["clayBots"] || newClay != checkpoints[newMinutes]["clay"] || obsBots != checkpoints[newMinutes]["obsBots"] || newObs != checkpoints[newMinutes]["obs"] || geodeBots != checkpoints[newMinutes]["geodeBots"] || newGeode != checkpoints[newMinutes]["geode"]) {
            return;
        }
    }
    
    if (oreBots >= checkpoints[newMinutes]["oreBots"] && newOre >= checkpoints[newMinutes]["ore"] && clayBots >= checkpoints[newMinutes]["clayBots"] && newClay >= checkpoints[newMinutes]["clay"] && obsBots >= checkpoints[newMinutes]["obsBots"] && newObs >= checkpoints[newMinutes]["obs"] && geodeBots >= checkpoints[newMinutes]["geodeBots"] && newGeode >= checkpoints[newMinutes]["geode"]) {
        // std::cout << newMinutes << " " << newGeode << " " << geodeBots << std::endl;
        checkpoints[newMinutes] = {{"oreBots", oreBots}, {"ore", newOre}, {"clayBots", clayBots}, {"clay", newClay}, {"obsBots", obsBots}, {"obs", newObs}, {"geodeBots", geodeBots}, {"geode", newGeode}};
    }

    runPrint(bestGeodes, blueprint, newMinutes, maxMinutes, oreBots, newOre, clayBots, newClay, obsBots, newObs, geodeBots, newGeode, maxOre, maxClay, maxObs);

    if (oreBuild) {
        runPrint(bestGeodes, blueprint, newMinutes, maxMinutes, oreBots + 1, newOre - blueprint["ore"], clayBots, newClay, obsBots, newObs, geodeBots, newGeode, maxOre, maxClay, maxObs);
    }

    if (clayBuild) {
        runPrint(bestGeodes, blueprint, newMinutes, maxMinutes, oreBots, newOre - blueprint["clay"], clayBots + 1, newClay, obsBots, newObs, geodeBots, newGeode, maxOre, maxClay, maxObs);
    }

    if (obsBuild) {
        runPrint(bestGeodes, blueprint, newMinutes, maxMinutes, oreBots, newOre - blueprint["obs-1"], clayBots, newClay - blueprint["obs-2"], obsBots + 1, newObs, geodeBots, newGeode, maxOre, maxClay, maxObs);
    }

    if (geodeBuild) {
        runPrint(bestGeodes, blueprint, newMinutes, maxMinutes, oreBots, newOre - blueprint["geode-1"], clayBots, newClay, obsBots, newObs - blueprint["geode-2"], geodeBots + 1, newGeode, maxOre, maxClay, maxObs);
    }
}

int main() {
    long result = 1;
    int qualitySum = 0;

    int bestGeodes = 0;
    int trueMax = 32;

    int maxOre = 4;
    int maxClay = 15;
    int maxObs = 15;

    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();

    std::cout << "Blueprint 1" << std::endl;
    std::map<std::string, int> blueprint1{{"ore", 2}, {"clay", 4}, {"obs-1", 4}, {"obs-2", 15}, {"geode-1", 2}, {"geode-2", 15}};
    for (int i = 1; i < trueMax; i++) {
        checkpoints[i] = {{"oreBots", 0}, {"ore", 0}, {"clayBots", 0}, {"clay", 0}, {"obsBots", 0}, {"obs", 0}, {"geodeBots", 0}, {"geode", 0}};
    }
    for (int maxMinutes=1; maxMinutes <= trueMax; maxMinutes++) {
        std::cout << "Minutes " << maxMinutes << std::endl;
        runPrint(bestGeodes, blueprint1, 0, maxMinutes, 1, 0, 0, 0, 0, 0, 0, 0, maxOre, maxClay, maxObs);
    }
    result = result * bestGeodes;
    qualitySum += 1 * bestGeodes;

    std::cout << "Blueprint 2" << std::endl;
    bestGeodes = 0;
    maxOre = 4;
    maxClay = 12;
    maxObs = 8;
    std::map<std::string, int> blueprint2{{"ore", 4}, {"clay", 4}, {"obs-1", 4}, {"obs-2", 12}, {"geode-1", 3}, {"geode-2", 8}};
    for (int i = 1; i < trueMax; i++) {
        checkpoints[i] = {{"oreBots", 0}, {"ore", 0}, {"clayBots", 0}, {"clay", 0}, {"obsBots", 0}, {"obs", 0}, {"geodeBots", 0}, {"geode", 0}};
    }
    for (int maxMinutes=1; maxMinutes <= trueMax; maxMinutes++) {
        std::cout << "Minutes " << maxMinutes << std::endl;
        runPrint(bestGeodes, blueprint2, 0, maxMinutes, 1, 0, 0, 0, 0, 0, 0, 0, maxOre, maxClay, maxObs);
    }
    result = result * bestGeodes;
    qualitySum += 2 * bestGeodes;

    std::cout << "Blueprint 3" << std::endl;
    bestGeodes = 0;
    maxOre = 4;
    maxClay = 17;
    maxObs = 16;
    std::map<std::string, int> blueprint3{{"ore", 4}, {"clay", 4}, {"obs-1", 4}, {"obs-2", 17}, {"geode-1", 4}, {"geode-2", 16}};
    for (int i = 1; i < trueMax; i++) {
        checkpoints[i] = {{"oreBots", 0}, {"ore", 0}, {"clayBots", 0}, {"clay", 0}, {"obsBots", 0}, {"obs", 0}, {"geodeBots", 0}, {"geode", 0}};
    }
    for (int maxMinutes=1; maxMinutes <= trueMax; maxMinutes++) {
        std::cout << "Minutes " << maxMinutes << std::endl;
        runPrint(bestGeodes, blueprint3, 0, maxMinutes, 1, 0, 0, 0, 0, 0, 0, 0, maxOre, maxClay, maxObs);
    }
    result = result * bestGeodes;
    qualitySum += 3 * bestGeodes;

    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
    std::cout << "Time difference = " << std::chrono::duration_cast<std::chrono::milliseconds>(end - begin).count() << "[ms]" << std::endl;

    std::cout << result << std::endl;
    std::cout << qualitySum << std::endl;
    return 0;
}

File: Day19.py
from copy import deepcopy
import math
import numpy as np
from utils import readFile, typeWrapper, Node, Dijkstra

data = readFile("input.txt")

blueprints = []
for line in data:
    blueprintData = line.split(" ")
    blueprint = {}
    blueprint["ore"] = int(blueprintData[6])
    blueprint["clay"] = int(blueprintData[12])
    blueprint["obs"] = (int(blueprintData[18]), int(blueprintData[21]))
    blueprint["geode"] = (int(blueprintData[27]), int(blueprintData[30]))
    blueprints.append(blueprint)

bestGeodes = 0
minuteMax = 26
checkpoints = {}

maxOre = None
maxClay = None
maxObs = None
def runPrint(blueprint, ore, clay, obs, geode, minutes):
    global bestGeodes
    global minuteMax
    global checkpoints

    global maxOre
    global maxClay
    global maxObs

    newMinutes = minutes + 1

    oreBuildable = ore[1] >= blueprint["ore"] and ore[0] < maxOre
    clayBuildable = ore[1] >= blueprint["clay"] and clay[0] < maxClay
    obsBuildable = ore[1] >= blueprint["obs"][0] and clay[1] >= blueprint["obs"][1] and obs[0] < maxObs
    geodeBuildable = ore[1] >= blueprint["geode"][0] and obs[1] >= blueprint["geode"][1]

    #Bot + Current Materials
    newOre = ore[1] + ore[0]
    newClay = clay[1] + clay[0]
    newObs = obs[1] + obs[0]
    newGeode = geode[1] + geode[0]

    if newMinutes >= minuteMax:
        if newGeode > bestGeodes:
            bestGeodes = newGeode
        return

    remainder = minuteMax - newMinutes
    maxPossible = newGeode + (remainder * geode[1]) + (remainder * (remainder + 1)) / 2
    # for newMin in range(1, remainder+1):
    #     maxPossible += geode[1] + newMin
    
    if maxPossible <= bestGeodes:
        return
    
    # for minute in range(1, newMinutes + 1):
    #     if ore[0] <= checkpoints[minute][0] and \
    #         clay[0] <= checkpoints[minute][1] and \
    #         obs[0] <= checkpoints[minute][2] and \
    #         geode[0] <= checkpoints[minute][3] and \
    #         newOre <= checkpoints[minute][4] and \
    #         newClay <= checkpoints[minute][5] and \
    #         newObs <= checkpoints[minute][6] and \
    #         newGeode <= checkpoints[minute][7]:
    #         return
    
    # if ore[0] >= checkpoints[newMinutes][0] and \
    #     clay[0] >= checkpoints[newMinutes][1] and \
    #     obs[0] >= checkpoints[newMinutes][2] and \
    #     geode[0] >= checkpoints[newMinutes][3] and \
    #     newOre >= checkpoints[newMinutes][4] and \
    #     newClay >= checkpoints[newMinutes][5] and \
    #     newObs >= checkpoints[newMinutes][6] and \
    #     newGeode >= checkpoints[newMinutes][7]:
    #     checkpoints[newMinutes] = (ore[0], clay[0], obs[0], geode[0], newOre, newClay, newObs, newGeode)

    runPrint(blueprint, (ore[0], newOre), (clay[0], newClay), (obs[0], newObs), (geode[0], newGeode), newMinutes)

    if oreBuildable:
        runPrint(blueprint, (ore[0] + 1, newOre - blueprint["ore"]), (clay[0], newClay), (obs[0], newObs), (geode[0], newGeode), newMinutes)

    if clayBuildable:
        runPrint(blueprint, (ore[0], newOre - blueprint["clay"]), (clay[0] + 1, newClay), (obs[0], newObs), (geode[0], newGeode), newMinutes)

    if obsBuildable:
        runPrint(blueprint, (ore[0], newOre - blueprint["obs"][0]), (clay[0], newClay - blueprint["obs"][1]), (obs[0] + 1, newObs), (geode[0], newGeode), newMinutes)

    if geodeBuildable:
        runPrint(blueprint, (ore[0], newOre - blueprint["geode"][0]), (clay[0], newClay), (obs[0], newObs - blueprint["geode"][1]), (geode[0] + 1, newGeode), newMinutes)

geodeProduct = 1
qualitySum = 0
for index in range(len(blueprints)):
    print(index + 1)
    blueprint = blueprints[index]
    maxOre = max(blueprint["ore"], blueprint["clay"], blueprint["obs"][0], blueprint["geode"][0])
    maxClay = blueprint["obs"][1]
    maxObs = blueprint["geode"][1]

    bestGeodes = 0
    runPrint(blueprint, (1, 0), (0, 0), (0, 0), (0, 0), 0)
    print(bestGeodes)
    geodeProduct *= bestGeodes
    qualitySum += (index + 1) * bestGeodes

print(geodeProduct)
print(qualitySum)